### ISSUE TITLE
dnf/rpm installs use user/group info from host machine

### ISSUE THREAD

**Issue by coling** [systemd/mkosi#805]
I mentioned this on the mailing list but figured I'd open a proper issue for more exposure.

So I didn't appreciate this before, but it seems a [long standing RPM issue](https://github.com/rpm-software-management/rpm/issues/882) where, when using --root the packages will be installed with the uid/gid mappings from the host machine rather than the passwd/group files from inside the root.

This makes for a problem using mkosi as it doesn't make the builds repeatable and very much dependant on the setup/configuration of the host machine.

For now, I've added the following to my mkosi.postinst:

  rpm -qa | xargs rpm --restore >/dev/null 2>&1

(for an RPM based target distro this is fine, but obviously others will presumably have similar commands if they too are affected. From the above bug, it appears Suse may patch RPM to always use user/group info from the root's /etc/passwd+group files)).

This work around however, isn't cheap. It takes ~2 minutes on a small package list on my semi-recent SSD laptop. That's pretty much the same time it takes to install the packages in the first place.

So my question is, is there a better work around for this? Or can this be somehow mitigated in mkosi?

While I don't think it helps, the user namespaces may be a useful work around (e.g. semi related to #716) but I think the real fix is likely in glibc/rpm. 

In some ways this is similar to #130 (dnf/rpm using host machine config)

**Comment by DaanDeMeyer** [comment]
This would likely be fixed by doing a multi stage build, where we first install dnf in a minimal chroot including /etc/passwd and then invoke dnf in that chroot using systemd-nspawn to install the final image.

We're looking to use the same approach to fix some of the other dnf/rpm issues we've been seeing, so this is another reason on top of those to try out this approach I guess

**Comment by DaanDeMeyer** [comment]
So I've changed my mind on this, rpm should simply be fixed to not use the user/group info from the host machine. I've already submitted a patch for dpkg that fixes the same thing there which got accepted, if I find time I'll do the same for RPM.

**Comment by DaanDeMeyer** [comment]
I realized we can do a clever trick to workaround the issue. We first install the package that provides /etc/passwd. Then, we run the package manager again to install the other packages. But this time, we mount /etc/passwd from the chroot on top of /etc/passwd from the host in a private mount namespace. This works around the issue as it doesn't matter anymore if the package manager uses /etc/passwd from the host as that will now point to /etc/passwd from the chroot. This was implemented in #1432 so I'm closing this issue.
